docs/topics/templates.txt

```diff
@@ -415,6 +415,22 @@ adds defaults that differ from Jinja2's for a few options:
 
   It defaults to an empty list.
 
+  Use context processors for non-trivial request-dependent values that you want
+  computed only once per template render (and not at all if no template is
+  rendered).
+
+  If a context processor's value isn't used in a majority of your templates or
+  isn't used more than once per template render, it would be more explicit to
+  put the function that you would use as a context processor in the template
+  context (for example, using ``jinja2.Environment`` as described below) and
+  then to call that function in the template::
+
+    {{ function(request) }}
+
+  Since context processors are run for every template that's rendered, think
+  carefully before putting expensive computations in them, especially if the
+  result isn't used in all or most of your templates.
+
```

Comment:
I like Aymeric's re-ordering to put the focus on the recommended approach.

For the cases where someone is reaching for a context processor (the context of these docs), I think middleware is never a preferable option (for the reasons @aaugustin lists), and best not mentioned here. The final comment about "not wanting to cache on a private attribute of the request" is more about the option that you could have a function which takes the request, and caches its result on the request to avoid repeat computation on repeat calls. I think this bullet point is confusing and should just be removed. Removing it does not imply that the three remaining bullet points can _only_ be satisfied by using a context processor, just that a context processor is one reasonable choice if those conditions apply. We don't need to get too deep into the weeds.

In my use case that drove adding context processors, really the motivating factor was verbosity of templates more than purity about caching on the request. We had a `me` variable, injected by a context processor, that was a JSON-serialized version of `request.user` (needed because these templates were shared by server and client, so all data referenced in them needed to be JSON serialized in a consistent way, not server-only Python objects). This `me` variable was used very frequently in the templates (usually accessing elements of it), and neither I nor the front-end team maintaining the templates relished the idea of modifying every occurrence of e.g. `{{ me.username }}` to something like `{{ me(request).username }}`.

But I agree that this is well into "matter of taste" territory.